### PR TITLE
Visual fingerprints in the CLI app (Issue #402)

### PR DESCRIPTION
**Pull request by deloachcd**
(Issue #402)

Hopefully this message finds you all well,
I implemented a terminal-color-code based system for displaying the visual fingerprints of the web browser extension and mobile app in the CLI in a simple fashion that does not require emoji support. This approach was created by reverse engineering the method utilized in the source code of the web and mobile app, and simply applying it in Python, with terminal color code sequences substituted for HTML color codes, and text strings substituted for icons.

Here's what the result looks like (the color codes could potentially be tweaked to look better):
demo

I've included test cases to ensure this feature's output is consistent with that of the web and mobile app (in terms of describing the icons, the colors are not one-to-one).
These tests were based off three references pictures, taken of the web extension's output:
password1
password2
password3

Of note is that this password entry method has you pressing "enter" to display the real fingerprint (rather than having a delay of 500ms), after which you can confirm by pressing "enter" again or make corrections with "backspace". This means that the user ends up pressing "enter" twice, rather than once.

This is my first pull request on a notable open source project, so any constructive criticism you may have would be greatly appreciated. Even if none of this code is actually integrated, I'm happy to have at least made an effort to contribute.

### CODE REVIEW

File: cli/lesspass/core.py [hunk 1]
```diff
@@ -30,7 +31,11 @@ def main(args=sys.argv[1:]):
         sys.exit(4)
 
     if not args.master_password:
-        args.master_password = getpass.getpass("Master Password: ")
+        prompt = "Master Password: "
+        if not args.fingerprint:
```
Comment (guillaumevincent) [inline cli/lesspass/core.py]: reverse your `if not else` into `if else`

**Comment by guillaumevincent** [comment]
Hello,
for a first one, this is a good one.
We will integrate this in LessPass.

What do you think about printing unicode string for all icons?

File: cli/lesspass/core.py (hunk 1 above)
Comment (deloachcd) [reply cli/lesspass/core.py]: resolved in 66b481c

**Comment by deloachcd** [comment]
that sounds like a good idea to me

**Comment by guillaumevincent** [comment]
I can't create a gist, but we could iterate over this file https://pastebin.com/kRcWfnMV 
I look for the icon here: http://www.fileformat.info

**Comment by deloachcd** [comment]
variable_width

Potential challenge: these unicode characters having variable widths can cause display issues when we're writing over the same 3 characters over and over again

**Comment by edouard-lopez** [comment]
Awesome! 
thanks for taking time to implement this, color-based approach is a really nice fallback idea!

**Comment by guillaumevincent** [comment]
@deloachcd are you waiting something from me?
We need to confirm all icons and colors are available.

**Comment by edouard-lopez** [comment]
I like the proposal, however I see 2 drawbacks in current version:
* user need to know what word is associated to each icon (assume user use UI 
app)
* it took me several attempts to (re)discover I need to type `Enter` to get the real fingerprint for my password. 

We need to either debounce before showing final password or a visual cue to tell user to type `Enter`:

	Master Password: birthday-cake      birthday-cake      certificate [Enter to get password]

I prefer a debounce mechanism.

**Comment by deloachcd** [comment]
@edouard-lopez I'll work on incorporating a debounce mechanism
@guillaumevincent I can work on this as well

I'll comment again once I've got a tentative solution working and committed

**Comment by deloachcd** [comment]
Update:
- debounce for "fake" password display as requested by @guillaumevincent 
has been implemented in 14c949c.
- translation of font awesome icons to unicode equivalents has been mapped out in https://gist.github.com/deloachcd/6bf1d00992af1bcf54dafd06104fa337

Unfortunately, not all font awesome icons are available in unicode as glyphs.
These have been marked in the gist with "TODO".

**Comment by edouard-lopez** [comment]
I made some suggestion for the unicode characters: https://gist.github.com/deloachcd/6bf1d00992af1bcf54dafd06104fa337#gistcomment-2984825

**Comment by deloachcd** [comment]
So it looks like after the suggestions @guillaumevincent provided, the remaining challenge is to determine whether or not the user can properly display unicode fonts in their terminal emulator. I'm not sure how to go about doing this, but I'll do some research into it when I have time over the next few days.

**Comment by deloachcd** [comment]
Update: by checking that the user is using a 256 color terminal and checking that [icons-in-terminal](https://github.com/sebastiencs/icons-in-terminal) is installed, we can get pretty close to how the webapp displays visual fingerprints:
render1564526543042

Without [icons-in-terminal](https://github.com/sebastiencs/icons-in-terminal) installed, however, it needs to fall back to plaintext.

If you are both OK with this solution, I can test what I have for it and commit it as part of this PR.

**Comment by edouard-lopez** [comment]
Screencast looks pretty good!

### Try to replicate

* installed `icons-in-terminal`
* edited [`user_has_unifont()`](https://github.com/lesspass/lesspass/pull/442/files#diff-ff1cbc5d08a828c56881709c51d506c3R14) to always return `True`
* got a messed up rendering

	lessspass-icons-fails

### ~/.config/fontconfig/conf.d/30-icons.conf

	<?xml version="1.0"?>
	<!DOCTYPE fontconfig SYSTEM "fonts.dtd">
	<fontconfig>
	  <alias><family>Noto Mono</family><prefer><family>icons-in-terminal</family></prefer></alias>
	</fontconfig>
	
### Icons

I can print icon font correctly

	~/apps/icons-in-terminal$ ./print_icons.sh
	
> image

**Comment by guillaumevincent** [comment]
Hello,
this is awesome, the first step can be to have icons-in-terminal installed. I don't know how hard it is, but the best will be to create lesspass.ttf font and merge it at runtime with the terminal font like  icons-in-terminal.

@deloachcd can we detect icons-in-terminal is installed and print an error message if not (windows users, and linux without the font)?

**Comment by deloachcd** [comment]
@edouard-lopez  try pulling in commit 2d1bf69 and let me know if it works for you
@guillaumevincent currently, it checks using `user_has_icons_in_terminal`:
```
def user_has_icons_in_terminal():
    return os.path.exists(os.path.expanduser("~/.fonts/icons-in-terminal.ttf"))
```
This ought to be a pretty reliable way of detected that the user has `icons-in-terminal` installed.
If this returns `False`, it falls back to using plaintext descriptions of the icons, ex. `hotel coffee database`

**Comment by guillaumevincent** [comment]
ping @edouard-lopez on this one

**Comment by ddelabru** [comment]
For what it's worth I have used a rebased version of this feature branch to install lesspass on my own system, because I was using the CLI version and really missed the fingerprint feature. It's pretty satisfying to use, especially combined with copying to clipboard -- neither my master password nor the generated password ever appears on screen, but I can tell that I entered my long master password correctly -- and I can accomplish that in the terminal environment.

I tried it with Kitty at first, but icons-in-terminal didn't consistently look right with that terminal emulator. (And it wasn't only a problem in lesspass; I saw that when I just tried the test scripts that came with icons-in-terminal, too.) I switched to GNOME Terminal for several reasons and lesspass CLI with fingerprints has worked quite well there. And if I temporarily move the icons-in-terminal font so the fingerprint doesn't use icons, it still looks pretty good.

Basically I'm going to keep installing from source so I can use this feature until it's merged :stuck_out_tongue:

**Comment by rstmsn** [comment]
@deloachcd your preview looks great, however in my terminal I'm finding the words somewhat difficult to read. Have you considered providing a foreground text colour, that provides good contrast based upon the particular background colour?

Screen Shot 2020-02-24 at 15 19 16

File: cli/lesspass/visual_fingerprint.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+import tty
+import termios
+import threading
+
+if os.name == "nt":
+    import msvcrt
+
+
+def user_has_icons_in_terminal():
+    return os.path.exists(os.path.expanduser("~/.fonts/icons-in-terminal.ttf"))
```
Comment (edouard-lopez) [inline cli/lesspass/visual_fingerprint.py:15]: We will need some documentation and either define limited targets platform or hardening

**Comment by guillaumevincent** [comment]
Thank you @deloachcd for this pull request

Sorry for the very long merge, but the installation of the icon font wasn't something I would like to merge.

I use icon unicode and enable it by default

Will see if the feedback is good

I also removed the color because it was not user friendly

I owe you a beer or two! :beers:
